### PR TITLE
Generate index.routing_path from dynamic mapping templates

### LINKED ISSUE #90528: Auto generate index.routing_path from dynamic mapping template.
This relates to #86790, which added logic that generates the `index.routing_path` index setting from from field mapped as keyword and marked as time series dimension in the mapping.

This builds on top of this and should also check dynamic templates if field template definition are of type `keyword` and `time_series_dimension` attribute is enabled. The `path_match` should then be added to the `index.routing_path` index setting.

This request originated from: https://github.com/elastic/kibana/issues/141796#issuecomment-1260619212

### PR DESCRIPTION
**Pull request by martijnvg**
Currently, when the `index.routing_path` index setting isn't specified, then it is auto generated from fields defined in the mapping.

This change also checks for dynamic templates with time series dimension keyword fields. The `path_match` attributes of these dynamic templates will then also be selected as routing path.

Closes #90528

### CODE REVIEW

**Review by nik9000 — APPROVED** [review]

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java [hunk 1]
```diff
@@ -161,19 +164,34 @@ private List<String> findRoutingPaths(String indexName, Settings allSettings, Li
...
-                        if (routingPaths == null) {
-                            routingPaths = new ArrayList<>();
-                        }
-                        routingPaths.add(keywordFieldMapper.name());
-                    }
-                }
+                routingPaths = extractPath(routingPaths, fieldMapper);
+            }
+            for (var template : mapperService.getAllDynamicTemplates()) {
+                var templateName = "__dynamic__" + template.name();
```
Comment (dakrone) [inline modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java:170]: Where does the `__dynamic__` come from? Is that a standard naming convention somewhere?

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java [hunk 2]
```diff
@@ -161,19 +164,34 @@ private List<String> findRoutingPaths(String indexName, Settings allSettings, Li
...
+                    .build(MapperBuilderContext.ROOT);
+                routingPaths = extractPath(routingPaths, mapper);
             }
             return routingPaths;
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
+    private static List<String> extractPath(List<String> routingPaths, Mapper mapper) {
```
Comment (dakrone) [inline modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java:185]: I really don't like that this method takes both a `routingPaths` parameter (which it mutates) and returns the `routingPaths` object. Can we make it either take `routingPaths` and mutate it but then return `void` (and add javadocs), or else make it not mutate the argument and return a new list?

**Review by dakrone — APPROVED** [review]
LGTM, I left two pretty minor comments

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java (hunk 1 above)
Comment (martijnvg) [reply modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java:170]: Not really a convention, but I the same template name being used when validating templates in `RootObjectMapper#validate(...)` method.

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java (hunk 2 above)
Comment (martijnvg) [reply modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java:185]: Prior to this change the `routingPaths` list was lazily initialised in the `findRoutingPaths(...)` method.
Similar logic was used for extracting paths from dynamic templates, so I generated this method.

Maybe it just easier to remove the lazy initialising of the list part and then the function can return void and update the provided list, which is your first suggestion.

**Pull request merged**